src/utils/AssayGeneratorMetabo.cpp

```diff
@@ -0,0 +1,523 @@
+    }
+
+    // annotate precursor mz and intensity
+    annotatePrecursorIntensity(spectra, 0.2, 0.2);
+
+    // filter feature by convexhull size
+    auto map_it = remove_if(feature_map.begin(), feature_map.end(),
+                                                 [&hull_size_filter](const Feature& f) -> bool
+                                                 {
+                                                   return f.getConvexHulls().size() < hull_size_filter;
+                                                 });
+    feature_map.erase(map_it, feature_map.end());
+
+    KDTreeFeatureMaps fp_map_kd;
+    vector<FeatureMap> v_fp;
+    v_fp.push_back(feature_map);
+    fp_map_kd.addMaps(v_fp);
+
+    // read FeatureMap in KDTree for feature-precursor assignment
+    map<const BaseFeature*, vector<size_t> > map_feature_to_precursors = extractMetaInformation(spectra, fp_map_kd, precursor_mz_tol, precursor_rt_tol);
+
+    std::vector<AssayRow> assaylib;
+    int transition_group_counter = 0;
+
+    for (std::map<const BaseFeature*, std::vector<size_t>>::iterator it = map_feature_to_precursors.begin();
+         it != map_feature_to_precursors.end();
+         ++it)
+    {
+
+      String description("UNKNOWN"), sumformula("UNKNOWN"), adduct("UNKNOWN");
+      const BaseFeature* min_distance_feature = it->first;
+
+      // extract metadata from featureXML
+      if (!(min_distance_feature->getPeptideIdentifications().empty()) &&
+          !(min_distance_feature->getPeptideIdentifications()[0].getHits().empty()))
+      {
+        description = min_distance_feature->getPeptideIdentifications()[0].getHits()[0].getMetaValue("description");
+        sumformula = min_distance_feature->getPeptideIdentifications()[0].getHits()[0].getMetaValue("chemical_formula");
+        adduct = min_distance_feature->getPeptideIdentifications()[0].getHits()[0].getMetaValue("modifications");
+
+        // change format of adduct information M+H;1+ -> [M+H]1+
+        String adduct_prefix = adduct.prefix(';').trim();
+        String adduct_suffix = adduct.suffix(';').trim();
+        adduct = "["+adduct_prefix+"]"+adduct_suffix;
+      }
+
+      // check if known unknown should be used
+      if (description == "UNKNOWN" && sumformula == "UNKNOWN" && adduct == "UNKNOWN" && !use_known_unknowns) { continue; }
+
+      double spectrum_rt = 0.0;
+      double highest_precursor_mz = 0.0;
+      float highest_precursor_int = 0.0;
+      MSSpectrum highest_precursor_int_spectrum;
+      MSSpectrum transition_spectrum;
+
+      // find precursor/spectrum with highest intensity precursor
+      std::vector<size_t> index = it->second;
+      for (std::vector<size_t>::iterator index_it = index.begin(); index_it != index.end(); ++index_it)
+      {
+        const MSSpectrum &spectrum = spectra[*index_it];
+        const vector<Precursor> &precursor = spectrum.getPrecursors();
+
+        // get m/z and intensity of precursor
+        double precursor_mz = precursor[0].getMZ();
```

Comment:
add a comment here that only spectra with precursors are in the map, therefore no need to check for presence of precursors